include/fmt/compile.h

```diff
@@ -536,15 +570,51 @@ template <typename T, typename Char> struct parse_specs_result {
   int next_arg_id;
 };
 
+constexpr int manual_indexing_id() { return -1; }
+
 template <typename T, typename Char>
 constexpr parse_specs_result<T, Char> parse_specs(basic_string_view<Char> str,
-                                                  size_t pos, int arg_id) {
+                                                  size_t pos, int next_arg_id) {
   str.remove_prefix(pos);
-  auto ctx = basic_format_parse_context<Char>(str, {}, arg_id + 1);
+  auto ctx = basic_format_parse_context<Char>(str, {}, next_arg_id);
   auto f = formatter<T, Char>();
   auto end = f.parse(ctx);
   return {f, pos + fmt::detail::to_unsigned(end - str.data()) + 1,
-          ctx.next_arg_id()};
+          next_arg_id == 0 ? manual_indexing_id() : ctx.next_arg_id()};
+}
+
+template <typename Char> struct arg_id_handler {
+  constexpr void on_error(const char* message) { throw format_error(message); }
+
+  constexpr int on_arg_id() {
+    throw format_error("handler cannot be used for empty arg_id");
+    return 0;
+  }
+
+  constexpr int on_arg_id(int id) {
+    arg_id = arg_ref<Char>(id);
+    return 0;
+  }
+
+  constexpr int on_arg_id(basic_string_view<Char> id) {
+    arg_id = arg_ref<Char>(id);
+    return 0;
+  }
+
+  arg_ref<Char> arg_id;
+};
+
+template <typename Char> struct parse_arg_id_result {
+  arg_ref<Char> arg_id;
+  const Char* arg_id_end;
+};
```

Comment:
Can we pass `begin` by reference in `parse_arg_id` and avoid introducing this struct?

Reply:
Hmm... it would be a reference to the pointer, or (_IMHO_ better) a pointer to the pointer, is it ok?

Reply:
Sure, I think reference is better unless it can be null.

Reply:
Actually, it's probably impossible because there is a need to have `arg_id_end` as a constexpr variable or, more importantly, begin has to be a non-constexpr variable in that case, but it should be used in constexpr context.